src/Clapi/NamespaceTracker.hs

```diff
@@ -145,113 +146,97 @@ liftedWaitThen onFwd onRev = do
-        let rejections = toReject ++ unsubErrors
-        when (not $ null $ rejections) $ lift . sendRev $ ServerData awu rejections
-        when (not $ null toFwd) $ lift . sendFwd $ ClientData awu toFwd'
-    fwd (ClientDisconnect awu) = do
-        rms <- handleClientDisconnect awu
-        lift . sendFwd $ ClientData awu rms
-        lift . sendFwd $ ClientDisconnect awu
-    rev :: (Ord i, Ord u, Monad m, Show i, Show u) => ServerEvent (AddrWithUser i u) [RoutableMessage] -> StateT
-        (Owners (AddrWithUser i u), Registered (AddrWithUser i u))
-        (NsProtocol m i u)
-        ()
-    rev (ServerData awu rms) = do
-        -- FIXME: is the fact that this looks like it's totally arsing up the
-        -- state to do with having to do the blasted contexts thing? Try
-        -- rewriting without using flippin' lenses:
-        stateFst $ updateOwnerships awu rms
-        -- FIXME: use of registereSubscriptions whined and whined about needing
-        -- FlexibleContexts, but I have not idea why...
-        stateSnd $ registerSubscriptions awu rms
-        fanOutBundle awu rms
-        stateSnd $ mapM_ handleDeletedNamespace rms
-    -- What do we do if we told the client to go away? Obviously pass that on
-    -- and drop registrations, but what about ownership?
-    -- Do they lose everything they owned?
-    rev (ServerDisconnect awu) =
-        lift . sendRev $ (ServerDisconnect awu)
-
--- Theoretically the left of the returned pair from this is a different type
--- since it should never contain an unsubscribe
+    sendErrorBundle i s ps = lift . sendRev $ ServerData i $
+        UpdateBundle (map (flip UMsgError s) ps) []
+    kick i = do
+        lift . sendRev $ ServerDisconnect i
+        fwd $ ClientDisconnect i
+    hasOwnership :: (Monad m, UMsg msg, Ord i, Show i) => i -> Ownership -> [msg] -> StateT (Owners i) m [Path]
+    hasOwnership i eo ms = get >>= \s -> return $ filter (\p -> Just eo == getPathOwnership i s p) $ map uMsgPath ms
+    fwd :: (Monad m, Ord i, Show i) => ClientEvent i Bundle x -> StateT (Owners i, Registered i) (NsProtocol m i) ()
+    fwd (ClientConnect i x) = return mempty
+    fwd (ClientData i b@(Left (UpdateBundle errs oums))) = do
+        badErrPaths <- stateFst $ hasOwnership i Client errs
+        badOumPaths <- stateFst $ hasOwnership i Client oums
+        if not $ null $ badErrPaths ++ badOumPaths
+            then sendErrorBundle i "Client role message in update bundle" $ allPaths b
+            else lift $ sendFwd ((i, Just errs), Request Nothing oums)
+    fwd (ClientData i (Right (RequestBundle subs dums))) = do
+        badSubPaths <- stateFst $ hasOwnership i Owner subs
+        badDumPaths <- stateFst $ hasOwnership i Owner dums
+        let allBadPaths = badSubPaths ++ badDumPaths
+        if not $ null $ allBadPaths
+            then do
+                sendErrorBundle i "Client role taken for own path" $ allBadPaths
```

Comment:
"Client role"? This message to me implies that we only have one client per path!

Reply:
Ok, I hadn't thought that role implied that you were the only one.
I'll try and reword.